Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -216,23 +249,41 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     /// Connection pools must be closed before they deinitialize.
     ///
     /// - returns: A future indicating close completion.
-    public func close() -> EventLoopFuture<Void> {
-        self.isClosed = true
-        return self.available.map { $0.close() }.flatten(on: self.eventLoop).map {
-            // inform any waiters that they will never be receiving a connection
-            while let waiter = self.waiters.popFirst() {
-                waiter.fail(ConnectionPoolError.closed)
+    public func shutdown() {
+        // synchronize access to closing
+        self.lock.lock()
+
+        // check to make sure we aren't double closing
+        guard !self.didShutdown else {
+            return
+        }
+        self.didShutdown = true
+
+        // is closed variable mutated
+        self.lock.unlock()
+
+        // no locks needed as this can only happen once
+        for available in self.available {
+            do {
+                try available.close().wait()
+            } catch {
+                #warning("TODO: use logger")
+                print("Could not close connection: \(error)")
             }
-            
-            // reset any variables to free up memory
-            self.available = []
-            self.activeConnections = 0
         }
+        // inform any waiters that they will never be receiving a connection
+        while let waiter = self.waiters.popFirst() {
```

Comment:
Could t this race with other accesses to the waiters array? Why not release the lock in a defer block right after acquiring it, so that it covers the entire method?

Reply:
I was thinking it shouldn't because the other methods will return immediately if `didShutdown` is true. But now that I think about it I do believe you are right.

Reply:
Yeah, I think it can race if another method is already running but past the `didShutdown` check.